passes/cmds/wrapcell.cc

```diff
@@ -54,6 +85,44 @@ std::optional<std::string> format(std::string fmt, const dict<IdString, Const> &
 	return {result.str()};
 }
 
+struct Chunk {
+	IdString port;
+	int base, len;
+
+	Chunk(IdString id, int base, int len)
+		: port(id), base(base), len(len) {}
+
+	IdString format(Cell *cell)
+	{
+		if (len == cell->getPort(port).size())
+			return port;
+		else if (len == 1)
+			return stringf("%s[%d]", port.c_str(), base);
+		else
+			return stringf("%s[%d:%d]", port.c_str(), base + len - 1, base);
+	}
+
+	SigSpec sample(Cell *cell)
+	{
+		return cell->getPort(port).extract(base, len);
+	}
+};
+
+std::vector<Chunk> collect_chunks(std::vector<std::pair<IdString, int>> bits)
```

Comment:
This deserves a comment or two about how it joins contiguous runs of bits. It's not a complex function but I'd rather not have to reverse engineer it to see that it's correct